components/script/dom/history.rs

```diff
@@ -86,17 +87,56 @@ impl History {
         // Step 5
         let serialized_data = StructuredCloneData::write(cx, data)?;
 
-        // TODO: Steps 6-7 Url Handling
-        // https://github.com/servo/servo/issues/19157
+        let new_url: ServoUrl = match url {
+            // Step 6
+            Some(urlstring) => {
+                let document_url = document.url();
+
+                // Step 6.1
+                let new_url = match ServoUrl::parse_with_base(Some(&document_url), &urlstring.0) {
+                    // Step 6.3
+                    Ok(parsed_url) => parsed_url,
+                    // Step 6.2
+                    Err(_) => return Err(Error::Security),
+                };
+
+                // Step 6.4
+                if new_url.scheme() != document_url.scheme() ||
+                   new_url.host() != document_url.host() ||
+                   new_url.port() != document_url.port() ||
+                   new_url.username() != document_url.username() ||
+                   new_url.password() != document_url.password()
+                {
+                    return Err(Error::Security);
+                }
+
+                // Step 6.5
+                if new_url.origin() != document_url.origin() {
+                    return Err(Error::Security);
+                }
+
+                new_url
+            },
+            // Step 7
+            None => {
+                let browsing_context_id = self.window.window_proxy().browsing_context_id();
```

Comment:
I think we should be able to get the URL from the document itself rather than going via the constellation.